src/Type.h

```diff
@@ -923,15 +923,10 @@ extern Type* flatten_type(Type* t);
 // Returns the "maximum" of two type tags, in a type-promotion sense.
 extern TypeTag max_type(TypeTag t1, TypeTag t2);
 
-// Given two types, returns the "merge", in which promotable types
-// are promoted to the maximum of the two.  Returns nil (and generates
-// an error message) if the types are incompatible.
-TypePtr merge_types(const TypePtr& t1, const TypePtr& t2);
```

Comment:
Unless vscode shows it wrong, this and `merge_type_list` below aren't in the `zeek::detail` namespace, so I believe technically they require a deprecation cycle and can be removed earliest in v6.1.

Practically, I'm unsure there are users.

Reply:
Yeah, I really doubt there are such users.  I only made `merge_types()` private to `Type.cc` to avoid the confusion of it being exposed but no one apparently using it - we could just keep it public if we want.

For `merge_type_list()`, I do think we should remove it (rather than fix its crash, which as I sketch would be a pain).  I can abide a deprecation cycle for that if the view is we should do so.

Reply:
Hmmmm `merge_type_list` takes an argument in the `detail::` namespace.  That seems contradictory ... ?

Reply:
> Hmmmm merge_type_list takes an argument in the detail:: namespace. That seems contradictory ... ?

That exists in other places as well (`Func::AddBody` / `Func::Invoke`). I would guess the "detail" object being passed/received from non-detail functions should mostly be treated as an opaque if possible.

I'm going to resolve this discussion assuming there are no out-of-tree users.

Reply:
> I only made merge_types() private to Type.cc to avoid the confusion of it being exposed but no one apparently using it - we could just keep it public if we want.I only made merge_types() private to Type.cc to avoid the confusion of it being exposed but no one apparently using it - we could just keep it public if we want.

> I'm going to resolve this discussion assuming there are no out-of-tree users.

Hmm, after more pondering. In the off-chance there being users of `merge_types()`, we don't actually provide a replacement that they could convert to. So I'm paddling back to keeping `merge_types()` available.

Reply:
I know it's tempting to call this unlikely and I indeed don't see any users of the function in our package collection, but our contract does actually say "Any changes to this code [i.e., the non-`detail` namespace] will strictly follow the Zeek team's deprecation process" (per the wiki). So I agree — we should either keep the function or put in a deprecation.

Reply:
Note, per the below, I've reverted the change to `merge_types()`.

Reply:
Ah — I read over the latest diff and was referring to `merge_type_list`. If I read the above right, we're in agreement that we should add deprecate it while introducing `maximal_type`, right?

Reply:
We can deprecate it, but we don't have a direct replacement.  `maximal_type` will work for simple (and IMHO "sane") cases, but it's not the full semantics.  Given that, what should the `[[deprecated("...")]]` message say?

Reply:
How about `[[deprecated("Remove in v6.1. Use maximal_type() if possible. See GH-2604.")]]`? That's similar to when we deprecated `MemoryAllocation()`, for which we also didn't provide a direct equivalent.

Reply:
Sounds good.  I've pushed a change to do this.